Azure/sonic-buildimage
# [201911][Monit] Restart telemetry container if memory usage is beyond the threshold

Signed-off-by: Yong Zhao <yozhao@microsoft.com>



#### Why I did it
This PR aims to monitor the memory usage of streaming telemetry container and restart streaming telemetry container if memory usage is larger than the pre-defined threshold.

#### How I did it
I borrowed the system tool Monit to run a script `memory_checker` which will periodically check the memory usage of streaming telemetry container. If the memory usage of telemetry container is larger than the pre-defined threshold for 10 times during 20 cycles, then an alerting message will be written into syslog and at the same time Monit will run the script `restart_container`  to restart the streaming telemetry container.

#### How to verify it
I verified this implementation on device `str-7260cx3-acs-1`.

    May 14 21:40:44.944347 str-7260cx3-acs-1 INFO systemd[1]: Starting LSB: service and resource monitoring daemon...
    May 14 21:40:44.956604 str-7260cx3-acs-1 INFO monit[15268]: Starting Monit 5.20.0 daemon with http interface at /var/run/monit.sock
    May 14 21:40:44.957122 str-7260cx3-acs-1 INFO monit[15268]: Monit start delay set to 300s
    May 14 21:40:44.957308 str-7260cx3-acs-1 INFO monit[15264]: Starting daemon monitor: monit.
    May 14 21:40:44.957823 str-7260cx3-acs-1 INFO systemd[1]: Started LSB: service and resource monitoring daemon.
    May 14 21:45:44.958178 str-7260cx3-acs-1 INFO monit[15270]: 'str-7260cx3-acs-1' Monit 5.20.0 started

    May 14 21:55:54.783286 str-7260cx3-acs-1 ERR monit[15270]: 'container_memory_telemetry' status failed (3) -- [telemetry]: Memory usage (600204902.4 Bytes) is larger than the threshold (419430400 Bytes)!
    May 14 21:55:54.783790 str-7260cx3-acs-1 INFO monit[15270]: 'container_memory_telemetry' exec: '/usr/bin/restart_container telemetry'
    May 14 21:55:54.995415 str-7260cx3-acs-1 INFO restart_container: Issue command 'sudo systemctl reset-failed telemetry.service' to reset failed status of service 'telemetry' ...
    May 14 21:55:55.171162 str-7260cx3-acs-1 INFO restart_container: Succeeded to reset failed status of service 'telemetry.service'.
    May 14 21:55:55.171756 str-7260cx3-acs-1 INFO restart_container: Issues command 'sudo systemctl restart telemetry.service' to restart container 'telemetry' ...
    May 14 21:55:55.556005 str-7260cx3-acs-1 INFO telemetry#supervisord 2021-05-14 21:55:55,555 WARN received SIGTERM indicating exit request
    May 14 21:55:55.562252 str-7260cx3-acs-1 INFO telemetry#supervisord 2021-05-14 21:55:55,559 INFO waiting for dialout, supervisor-proc-exit-listener, rsyslogd to die
    May 14 21:55:55.603269 str-7260cx3-acs-1 INFO telemetry#supervisord 2021-05-14 21:55:55,596 INFO stopped: dialout (terminated by SIGTERM)
    May 14 21:55:58.255612 str-7260cx3-acs-1 INFO telemetry.sh[8335]: 0
    May 14 21:55:58.528473 str-7260cx3-acs-1 INFO telemetry.sh[18723]: Starting existing telemetry container with HWSKU Arista-7260CX3-D108C8
    May 14 21:55:58.996912 str-7260cx3-acs-1 INFO restart_container: Issues command 'docker inspect -f \{\{.State.Running\}\} telemetry' to check running status of container 'telemetry' ...
    May 14 21:55:59.803825 str-7260cx3-acs-1 INFO restart_container: Container 'telemetry' is restarted successfully and is in the running state.

#### Which release branch to backport (provide reason below if selected)



- [x ] 201911

#### Description for the changelog



#### A picture of a cute animal (not mandatory but encouraged)
